Proyecto Métodos Númericos 

Profesor: Joaquín Peña Acevedo

Alumno: Ricardo Alberto Gloria Picazzo

# <center> Cuadratura Gaussiana para integrales dobles y triples </center>

## Introducción 

A lo largo de este proyecto se desarrolla la idea para aplicar la Cuadratura Gaussiana a integrales dobles y triples. Se implementan dos algortimos que aproximan integrales dobles y triples respectivamente usando la cuadratura Gaussiana. Se agregan tres ejemplos para cada algortimo con el proposito de probar las implementaciones. Dichos algoritmos fueron obtenidos del libro "Numerical Analysis" de Burden y Douglas. Antes de dichas implementaciones se agregó una explicación de la Cuadratura Gaussiana. También se desarrollan las propiedades de los polinomios de Legendre y su importancia para la Cuadratura Gaussiana. Se explica también cómo generalizar la cuadratura Gaussiana para regiones distintas a rectángulos de intervalos $[-1, 1]$. Finalmente se agragan las conclusiones del proyecto y las referencias. 

## Cuadratura Gaussiana

Recordamos que el método basico para aproximar integrales el la cuadratura númerica que consiste en aproximar $\int_a^bf$ con una suma de la forma $\sum_{i=0}^n a_i f(x_i)$. Así, la precisión de nuestra aproximación depende en la elección de los nodos $\{ x_0, x_1, \ldots, x_n\}$ y los coeficientes $\{c_0, c_1, \ldots, c_n\}$. Considerar nodos equiespaciados nos permite contar con formulas sencillas y fáciles de evaluar. Sin embargo, no nos asegura contar con una buena aproximación. La cuadratura Gaussiana se basa en elegir nodos $\{x_1, \ldots, x_n\}\subset [a, b]$ y coeficientes $c_1, c_2, \ldots, c_n$ de manera de minimizar el error en la aproximación $\sum_{i=1}^n c_if(x_i)$. Debemos elegir un total de $2n$ parametros: los nodos y los coeficientes. Los polinomios de grado igual o menor que $2n-1$ también tienen $2n$ parámetros. Por lo tanto el objetivo es que la suma $\sum_{i=1}^n c_if(x_i)$ arroje el valor exacto de la integral $\int_{-1}^1 P$ donde $P$ es un pololinomio de grado igual o menor que $2n-1$. 


### Polinomios de Legendre

Recordemos que el espacio de funciones continuas $C^0([-1, 1], \mathbb{R})$ es un espacio vectorial donde podemos definir el producto interno $\langle f, g \rangle := \int_{-1}^1 fg$. Recordamos también que dos vectores son ortogonales si $\langle f, g\rangle = 0$. Un conjunto ortogonal que es importante para el problema son los Polinomios de Legendre $\{P_n\}_{n\in\mathbb{N}}$ donde cada $P_n$ es de grado $n$. Dichos polinomios satisfacen las siguientes dos propiedades:

1. Para cada $n$, se tiene que $P_n$ es un polinomio mónico.
2. $\langle P, P_n \rangle = \int_{-1}^1 PP_n = 0$ siempre que $P$ sea de grado menor a $n$. 

Además, las raices de los Polinomios de Legendre son distintas, están contenidas es $(-1, 1)$ y son símetricas con respecto a $0$. Lo más importante es que satisfacen el

**Teorema.** Sean $x_1, \ldots, x_n$ las raíces del $n$-ésimo polinomio de Legendre $P_n(x)$. Para cada $i=1, 2, \ldots, n$ definimos

\begin{align*}
    c_i &:= \int_{-1}^1 \prod_{j=1\\j\neq i}^n \frac{x-x_j}{x_i-x_j} \, dx. \quad \quad\quad \quad (1)
\end{align*}
Si $P(x)$ es un polinomio de grado menor a $2n$, entonces
\begin{align*}
    \int_{-1}^1 P(x) \, dx = \sum_{i=1}^n c_iP(x_i).
\end{align*}

### Intervalos arbitrarios

Los polinomios de Legendre nos permiten aproximar integrales de la manera deseada pero sólo en el intervalo $[-1, 1]$. Sin embargo este inconveniente se puede solucionar con el Teorema de cambio de variable. Así, si deseamos estimar la integral de $f$ sobre $[a, b]$ podemos usar el cambio de variable:
\begin{align*}
    t = \frac{2x -a - b}{b-a} \iff x = \frac{(b-1)t + a +b}{2}
\end{align*}
Entonces 
\begin{align*}
    \int_a^b f(x) dx = \frac{b-a}{2}\int_{-1}^1 f\left(\frac{(b-1)t + (b+a)}{2} \right) \, dt.   \quad\quad\quad(2)
\end{align*}

## Integrales dobles

Ya conocemos la idea de como aproximar de una muy buena manera la integral de un función $f:[a, b] \to \mathbb{R}$ utilizando la Cuadratura Gaussiana. Con el uso del Teorema de Fubini, podemos utilizar lo anterior para aproximar integrales de funciones $f: \Omega \to \mathbb{R}$ donde $\Omega = \{(x, y)\in \mathbb{R}^2: a\leq x \leq b, c(x) \leq y \leq d(x)\}$ y $c, d$ son funciones continuas. Esto es calcularemos 
\begin{align*}
    \int_a^b \int_{c(x)}^{d(x)} f(x, y) \, dy \, dx.
\end{align*}
Para ello se requiere transformar, para cada $x\in [a, b]$, la variable $y\in [c(x, d(x)]$ en una variable $t\in [-1, 1]$. Usando la transformación lineal:

\begin{align*}
    f(x, y) &= f\left( x, \frac{(d(x) - c(x))t + d(x) + c(x)}{2}  \right),\\
    dy &= \frac{d(x) - c(x)}{2}\, dt.
\end{align*}

Así, para cada $x\in [a, b]$ se obtiene:
\begin{align*}
    \int_{c(x)}^{d(x)} f(x, y) \, dy = \frac{d(x)-c(x)}{2}\int_{-1}^1 f\left(x, \frac{(d(x)-c(x))t + d(x) + c(x)}{2}\right) \, dt.
\end{align*}

Luego, 
\begin{align*}
    \int_a^b \int_{c(x)}^{d(x)} f(x, y) \, dy\, dx \approx \int_a^b \frac{d(x)-c(x)}{2} \sum_{j=1}^n c_{n, j}f\left(x, \frac{(d(x) - c(x))r_{n, j} + d(x) + c(x)}{2} \right)   \, dx \quad\quad\quad (3)
\end{align*}
donde $r_{n, j}$ es la $j$-ésima raíz del polinomio de Legendre de grado $n$ y $c_{n, j}$ es el $j$-ésimo coeficiente definido en (1) para el polinomio de Legendre de grado $n$. 

Procedemos a implementar la integral Gaussiana doble:


In [2]:
import numpy as np
from scipy.special import roots_legendre

# Los parametros m y n determinan el número de particiones para cada eje
def GaussianDoubleIntegral(f, a, b, c, d, m, n):
    #importar los valores las raices y coeficientes de los polinomios de Legendre 
    r_m, c_m = roots_legendre(m, mu = False) 
    r_n, c_n = roots_legendre(n, mu = False)
    
    h1 = (b-a)/2
    h2 = (b+a)/2
    J = 0
    for i in range(m):
        JX = 0
        x = h1*r_m[i] + h2
        d1 = d(x)
        c1 = c(x)
        k1 = (d1 - c1)/2
        k2 = (d1 + c1)/2
        for j in range(n):
            y = k1*r_n[j] + k2 
            Q = f(x, y)
            JX += c_n[j]*Q
        J += c_m[i]*k1*JX 
        
    J = h1*J
    return J

Probemos la función de la integral doble implementada anteriormente.  

### Ejemplo 1.1

Calculemos la integral:

\begin{align*}
    \int_{0.1}^{0.5} \int_{x^3}^{x^2} e^{y/x} dy dx 
    &= \int_{0.1}^{0.5} \left( xe^{y/x}\right)\bigg|_{y=x^3}^{x^2}\, dx \\
    &= \int_{0.1}^{0.5} x(e^x - e^{x^2}) dx\\
    &= \left( e^x(x-1)-\frac{e^{x^2}}{2}  \right)\bigg|_{0.1}^{0.5}.
\end{align*}

Evaluamos para ver el valor de la integral anterior:


In [3]:
def EvalAntiderivada(x):
    r1 = np.exp(x)*(x-1)
    r2 = np.exp(x**2)/2
    return r1 - r2
Ia = EvalAntiderivada(0.5) - EvalAntiderivada(0.1)
print("El valor de la integral análitica es: ", Ia)

El valor de la integral análitica es:  0.033305566116232255


Así, 
\begin{align*}
    \int_{0.1}^{0.5} \int_{x^3}^{x^2} e^{y/x} dy dx  = 0.033305566116232255.
\end{align*}

Calculemos ahora la integral de manera númerica con la función anterior:



In [4]:
def f(x, y):
    return np.exp(y/x)
def c(x):
    return x**3
def d(x):
    return x**2

Inum = GaussianDoubleIntegral(f, 0.1, 0.5, c, d, 5, 5)    
print("La evaluación númerica de la integral es: ", Inum)
print("Diferencia de la integral análitica y la númerica: ", np.abs(Ia - Inum))

La evaluación númerica de la integral es:  0.033305566118675335
Diferencia de la integral análitica y la númerica:  2.4430804601571765e-12


Podemos ver que la integral númerica aproxima al valor de la integral 11 lugares decimales con sólo $mn = 25$ evaluaciones de la función. En Numerical Analysis de Burden y Faires se menciona que, para la integral anterior, con el método de Simpson para integrales dobles se obtiene una precisión de $7$ lugares decimales con $m=n=10$. Al usar $n=m=10$, se evalua la función $121$ veces.

### Ejemplo 1.2

Veamos ahora la integral:

\begin{align*}
    \int_{-1}^1 \int_{-2|x|}^{|x|} e^{x+y} \, dy \, dx
     &= \int_{-1}^1 (e^{x + |x|} - e^{x - 2|x|} ) \, dx\\
     &= \int_{-1}^0 (1 - e^{3x}) \, dx + \int_0^1 (e^{2x} - e^{-x})\, dx\\
     &= \left( x - \frac{e^{3x}}{3} \right)\bigg|_{-1}^{0} + \left( \frac{e^{2x}}{2} + e^{-x} \right)\bigg|_{0}^1\\
     &= \frac{1}{3e^3} + \frac{1}{e} + \frac{e^2}{2} - \frac{5}{6}.
\end{align*}

Evaluamos la integral análitica:


In [5]:
Ia = (1/3)*np.exp(-3) + np.exp(-1) + np.exp(2)/2 - 5/6
print("El valor de la integral análitica es: ", Ia)

El valor de la integral análitica es:  3.2456698467593887


Así, 
\begin{align*}
    \int_{-1}^1 \int_{-2|x|}^{|x|} e^{x+y} \, dy \, dx = 3.2456698467593887.
\end{align*}

Calculamos la integral de manera númerica:

In [6]:
def f(x, y):
    return np.exp(x+y)

def c(x):
    return -2*np.abs(x)

def d(x):
    return np.abs(x)

a = -1
b = 1
Inum = GaussianDoubleIntegral(f, a, b, c, d, 30, 30)
print("La evaluación númerica de la integral es: ", Inum)
print("Diferencia de la integral análitica y la númerica: ", np.abs(Ia - Inum))

La evaluación númerica de la integral es:  3.2483198878694486
Diferencia de la integral análitica y la númerica:  0.002650041110059842


Notése que en este caso se obtiene una menor precisión que en el ejemplo anterior aún usando $n=m=30$. Esto se puede deber a que la región de integración en el ejemplo 1 es muy pequeño a comparación de la región de integración de éste ejemplo.  

### Ejemplo 1.3

En este ejemplo integraremos una función $f:\mathbb{R}^2 \to \mathbb{R}$ donde para cada $x$ fijo, $f(x, y)$ es un polinomio de grado $4$. Usaremos $n=m=6$. Calculemos análiticamente la integral:
\begin{align*}
    \int_{-1}^1 \int_{-1}^1 21x^2y^6 + 6x^3y^5 + 16y^3 + 6x^3y^2 + 4y + 5 \, dy \, dx
    &= \int_{-1}^1 \left( 3x^2y^7 + x^3y^6 + 4y^4 + 2x^3y^3 + 2y^2 + 5y \right)\bigg|_{y=-1}^1 dx\\
    &= \int_{-1}^1 6x^2 + 4x^3 + 10 \, dx\\
    &= ( 2x^3 + x^4 + 10x)\bigg|_{-1}^1 \\
    &= 2 + 1 + 10 + 2 - 1 + 10 \\
    &= 24.
\end{align*} 

Ahora evaluemos la integral de manera númerica:

In [14]:
def f(x, y):
    r1 = (21*(x**2)*(y**6) + 6*(x**3)*(y**5) + 16*(y**3))
    r2 = (6*(x**3)*(y**2) + 4*y + 5)
    return r1 + r2

def c(x):
    return -1

def d(x):
    return 1

a = -1
b = 1
Inum = GaussianDoubleIntegral(f, a, b, c, d, 5, 5)
print("La evaluación númerica de la integral es: ", Inum)



La evaluación númerica de la integral es:  24.0


Observamos que en este caso la evaluación númerica nos da el valor exacto de la integral. Recordemos que la cuadratura de Gauss nos da el valor exacto de la integral para polinomos de grado igual o menor a $2n-1$. La función $f$ es un polinomio de grado $6$ para cada $x$ fijo. Es decir la integral interior de la integral doble es un polinomio de grado $6$. Al integrar para cada $x$ fijo se obtiene otra integral (la exterior) del polinomio $6x^2 + 4x^3 + 10$, el cual es de grado $2$. Así, al usar $n=m=5$ se obtiene el valor exacto de la integral en sólo $25$ evaluaciones de la integral. 

## Integrales triples

Nuevamente podemos aprovechar lo desarrollado para integrales dobles para aproximar integrales triples. De manera similar a como escribimos (2) y (3) podemos obtener mediante un cambio de varible: 
\begin{align*}
    \int_a^b \int_{c(x)}^{d(x)} \int_{\alpha(x, y)}^{\beta(x, y)}f(x, y, z) \, dz \, dy\, dx \approx \int_a^b \int_{c(x)}^{d(x)}\frac{\beta(x)-\alpha(x)}{2} \sum_{j=1}^n c_{n, j}f\left(x, y, \frac{(\beta(x, y) - \alpha(x, y))r_{n, j} + \beta(x, y) + \alpha(x, y)}{2} \right) \, dy \, dx 
\end{align*}

Implementemos así, la Integral Gaussiana Triple. 

In [8]:
# Los parametros m, n y p determinan el número de particiones para cada eje
def GaussianTripleIntegral(f, a, b, c, d, alpha, beta, m, n, p):
    #importar los valores las raices y coeficientes de los polinomios de Legendre 
    r_m, c_m = roots_legendre(m, mu = False)
    r_n, c_n = roots_legendre(n, mu = False)
    r_p, c_p = roots_legendre(p, mu = False)
    
    h1 = (b - a)/2
    h2 = (b + a)/2
    J = 0
    for i in range(m):
        JX = 0
        x = h1*r_m[i] + h2
        d1 = d(x)
        c1 = c(x)
        k1 = (d1 - c1)/2
        k2 = (d1 + c1)/2
        for j in range(n):
            JY = 0
            y = k1*r_n[j] + k2
            beta1 = beta(x, y)
            alpha1 = alpha(x, y)
            l1 = (beta1 - alpha1)/2
            l2 = (beta1 + alpha1)/2
            for k in range(p):
                z = l1*r_p[k] + l2
                Q = f(x, y, z)
                JY += c_p[k]*Q   
            JX += c_n[j]*l1*JY   
        J += c_m[i]*k1*JX        
    J = h1*J
    return J

### Ejemplo 2.1 


Calcularemos el volumen de la bola unitaria cetrada en el origen $S\subset \mathbb{R}^3$. El volumen de $S$ se puede calcular mediante la integral:
\begin{align*}
    vol(S) := \int_S 1 = \int_{-1}^1 \int_{-\sqrt{1-x^2}}^{\sqrt{1-x^2}} \int_{-\sqrt{1-x^2-y^2}}^{\sqrt{1-x^2-y^2}} 1 \, dz\, dy \, dy.
\end{align*}
La integral anterior es un tanto complicada de resolver de manera análitica. Eváluemosla de manera análitica con la implementación de la cuadratura Gaussiana pa integrales triples.


In [9]:
def f(x, y, z):
    return 1

def c(x):
    return -np.sqrt(1-x**2)

def d(x):
    return np.sqrt(1-x**2)

def alpha(x, y):
    return -np.sqrt(1 - x**2 - y**2)

def beta(x, y):
    return np.sqrt(1 - x**2 - y**2)

a = -1
b = 1

m = 50
n = m
p = m

Vol1 = GaussianTripleIntegral(f, a, b, c, d, alpha, beta, m, n, p)
print("El volumen aproximado de la esfera es: ", Vol1)


El volumen aproximado de la esfera es:  4.188807265415295


Para calcular $vol(S)$ de manera análitica podemos utilizar el Teorema de Cambio de Variable con las coordenadas cilídricas $\phi(r, \theta, \psi)=(r\sin(\psi)\cos(\theta), r\sin(\psi)\sin(\theta), r\cos(\psi))$ cuyo jacobiano es $J\phi = r^2\sin(\psi)$. Así:

\begin{align}
    vol(S) &:= \int_S 1 = \int_0^1 \int_0^{2\pi} \int_0^\pi r^2\sin(\psi) \, d\psi \, d\theta \, dr \quad\quad(1)\\
    &= \int_0^1 r^2 \int_0^{2\pi} \int_0^\pi \sin(\psi) \, d\psi \, d\theta \, dr\\
    &= 2\int_0^1 r^2 \int_0^{2\pi}  \, d\theta \, dr\\
    &= 4\pi \int_0^1  r^2 \, dr\\
    &= \frac{4\pi}{3} 
\end{align}

Evaluamos $vol(S)=\frac{4\pi}{3}$ y comparamos con la aproximación antes obtenida:

In [10]:
Vol = (4*np.pi)/3
print("Vol(S) = ", Vol)
print("Error de la aproximación: ", np.abs(Vol - Vol1))

Vol(S) =  4.1887902047863905
Error de la aproximación:  1.7060628904275177e-05


Se obtuvo una presición de casi 4 cifras decimales utilizando $m=n=p=50$. Resolvamos ahora de manera númerica la integral descrita en (1).

In [11]:
def f(r, theta, psi):
    return (r**2)*np.sin(psi)

def c(r):
    return 0

def d(r):
    return 2*np.pi

def alpha(r, theta):
    return 0

def beta(r, theta):
    return np.pi

a = 0
b = 1

m = 50
n = m
p = m

Vol2 = GaussianTripleIntegral(f, a, b, c, d, alpha, beta, m, n, p)
print("El volumen aproximado de la esfera es: ", Vol2)
print("Error de la aproximación: ", np.abs(Vol - Vol2))


El volumen aproximado de la esfera es:  4.1887902047863985
Error de la aproximación:  7.993605777301127e-15


Con el planteamiteto usando coordenadas esfericas se obtuvo una muy buena presisión de 14 cifras decimales. 

## Ejemplo 2.2

Veamos un análogo para $\mathbb{R}^3$ del Ejemplo 1.3. Probemos la implementación de la cuadratura de Gauss para integrales triples con una función $f:\mathbb{R}^3\to \mathbb{R}$ tal que fijando cada par de variables, $f$ es un polinomio de grado menor o igual a 4. Resolvamos así la integral:
\begin{align*}
    \int_{-1}^1 \int_{-1}^1 \int_{-1}^1 (8xyz - 60x^5y^4z + 4z - 14yz + 2x + 10) \, dz\, dy \, dx
    &= \int_{-1}^1\int_{-1}^1 (4xyz^2 - 30x^5y^4z^2 + 2z^2 - 7yz + 2xz + 10z)\bigg|_{-1}^1 \, dy\, dx\\
    &= \int_{-1}^1 \int_{-1}^1 (-14y + 4x + 20) \, dy\, dx\\
    &= \int_{-1}^1 (-7y^2 + 4xy + 20y)\bigg|_{-1}^1 \, dx\\
    &= \int_{-1}^1 (8x + 40) \, dx\\
    &= (4x^2 + 40x)\bigg|_{-1}^1 \\
    &= 80.
\end{align*}

Esperamos que el algoritmo arroje el resultado exacto de la integral:

In [12]:
def f(x, y, z):
    return (8*x*y*z - 60*(x**5)*(y**4)*z - 4*z - 14*y*z + 2*x + 10)

def c(x):
    return -1

def d(x):
    return 1

def alpha(x, y):
    return -1

def beta(x, y):
    return 1

a = -1
b = 1

m = 7
n = m
p = m

I = GaussianTripleIntegral(f, a, b, c, d, alpha, beta, m, n, p)
print("El valor de la integral es: ", I)



El valor de la integral es:  80.0


### Ejemplo 2.3

Recordemos que el centro de masa de un solido $S\subset \mathbb{R}$ con función de densidad $\delta$ es $(\overline{x}, \overline{y}, \overline{z})$ donde
\begin{align*}
    \overline{x} := \frac{\int_S \delta x}{\int_S \delta}, \quad
    \overline{y} := \frac{\int_S \delta y}{\int_S \delta},\quad\text{y}\quad
    \overline{z} := \frac{\int_S \delta z}{\int_S \delta}. 
\end{align*}

Calcularemos el centro de masa del cono $S\subset\mathbb{R}^3$ dado por $z^2 = x^2 + y^2$ y acotado por el plano $z=2$. ![title](f1.png)


Supondremos que el $S$ tiene una función de densidad de $\delta(x, y, z) = \sqrt{x^2 + y^2}$. 

Primero calcularemos el centro de masa de manera análitica utilizando el Teorema de cambio de variable con las coordenadas cilídricas $\phi(r, \theta, z) = (r\cos(\theta), r\sin(\theta), z)$. Recordamos que el jacobiano de $\phi$ es $J\phi=r$. Se tiene además que $\delta(x, y, z) = \sqrt{x^2+y^2} = r$. Así,

\begin{align*} 
    \int_S \delta &= \int_0^2 \int_0^z \int_0^{2\pi} r^2 \, d\theta \, dr\, dz
    = 2\pi \int_0^2 \int_0^z r^2 \, dr\, dz
    = \frac{2\pi}{3} \int_0^2 z^3 \, dz
    = \frac{2\pi}{12} (16) = \frac{8\pi}{3},\\
    \int_S \delta x &= \int_0^2 \int_0^z \int_0^{2\pi} r^2 \cos(\theta) \, d\theta\, dr\, dz 
    = \int_0^2 \int_0^z r^2 \int_0^{2\pi}  \cos(\theta) \, d\theta\, dr\, dz 
    = \int_0^2 \int_0^z r^2 (0) \, dr\, dz = 0,\\
    \int_S \delta y &= \int_0^2 \int_0^z \int_0^{2\pi} r^2 \sin(\theta) \, d\theta\, dr\, dz 
    = \int_0^2 \int_0^z r^2 \int_0^{2\pi}  \sin(\theta) \, d\theta\, dr\, dz 
    = \int_0^2 \int_0^z r^2 (0) \, dr\, dz = 0,\\
    \int_S \delta z &= \int_0^2\int_0^z\int_0^{2\pi} r^2 z \, d\theta \, dr \, dz 
    = 2\pi \int_0^2 z\int_0^z r^2 \, dr\, dz 
    = \frac{2\pi}{3} \int_0^2 r^4 \, dr = \frac{2\pi}{15}(2^5) = \frac{64\pi}{15}.
\end{align*}

Así, el centro de masa de $S$ es exactamente
\begin{align*}
    (\overline{x}, \overline{y}, \overline{z}) = \left( 0, 0, \frac{64\pi\cdot 3}{15\cdot 8\pi} \right) =  \left(0, 0, \frac{8}{5} \right) = (0, 0, 1.6).
\end{align*}

Supongamos que no conocemos el Teorema del Cambio de Variable. Así, las integrales anteriores se vuelven un tanto complicadas de evaluar:
\begin{align*}
    \int_S \delta &= \int_{-2}^2 \int_{-\sqrt{4-x^2}}^{\sqrt{4-x^2}} \int_{\sqrt{x^2 + y^2}}^2 \sqrt{x^2 + y^2} \, dz\, dy\, dx, \\
    \int_S \delta x &= \int_{-2}^2 \int_{-\sqrt{4-x^2}}^{\sqrt{4-x^2}} \int_{\sqrt{x^2 + y^2}}^2 x\sqrt{x^2 + y^2} \, dz\, dy\, dx, \\
    \int_S \delta y &= \int_{-2}^2 \int_{-\sqrt{4-x^2}}^{\sqrt{4-x^2}} \int_{\sqrt{x^2 + y^2}}^2 y\sqrt{x^2 + y^2} \, dz\, dy\, dx, \\
    \int_S \delta z &= \int_{-2}^2 \int_{-\sqrt{4-x^2}}^{\sqrt{4-x^2}} \int_{\sqrt{x^2 + y^2}}^2 z\sqrt{x^2 + y^2} \, dz\, dy\, dx.
\end{align*}

Así, necesitamos evaluar las integrales anteriores de manera númerica.


In [13]:
def delta(x, y, z):
    r = np.sqrt(x**2 + y**2)
    return r

def delta_x(x, y, z):
    r = np.sqrt(x**2 + y**2)
    return x*r

def delta_y(x, y, z):
    r = np.sqrt(x**2 + y**2)
    return y*r

def delta_z(x, y, z):
    r = np.sqrt(x**2 + y**2)
    return z*r

def c(x):
    r = np.sqrt(4 - x**2)
    return -r

def d(x):
    r = np.sqrt(4 - x**2)
    return r

def alpha(x, y):
    return np.sqrt(x**2 + y**2)

def beta(x, y):
    return 2

a = -2
b = 2

m = 5
n = m
p = m

I = GaussianTripleIntegral(delta, a, b, c, d, alpha, beta, m, n, p)
I_yz = GaussianTripleIntegral(delta_x, a, b, c, d, alpha, beta, m, n, p)
I_xz = GaussianTripleIntegral(delta_y, a, b, c, d, alpha, beta, m, n, p)
I_xy = GaussianTripleIntegral(delta_z, a, b, c, d, alpha, beta, m, n, p)

print("El centro de masa obtenido es (x, y, z) donde: ")
x = I_yz/I
y = I_xz/I
z = I_xy/I
print("x = ", x)
print("y = ", y)
print("z = ", z)

centro_masa = np.array([x, y, z])
exacto = np.array([0, 0, 1.6])

print("Error:\n\tLa distancia entre ambos centros de masa es de ", np.linalg.norm(centro_masa - exacto))

El centro de masa obtenido es (x, y, z) donde: 
x =  0.0
y =  1.1671287553528939e-17
z =  1.6495126045634954
Error:
	La distancia entre ambos centros de masa es de  0.049512604563495266


Las cuatro integrales fueron evaludas con $m=n=p=5$, lo cual implica la evaluación de 125 de cada integrando. El valor que arroja el algoritmo para $\overline{x}$ es el exacto. Mientras que para $\overline{y}$ se obtiene un error muy pequeño. Sin embargo, para $\overline{z}$ se obtiene sólo una exactitud de dos cifras decimales.

## Conclusiones
El método de la Cuadratura Gaussiana nos permite aproximar funciones de una manera precisa. Además al ser un método númerico de cuadratura, la implementación del algoritmo es sencilla. Al ser un método preciso, es también un método eficiente por nos brinda una muy buena aproximación sin la necesidad de evaluar el integrando una gran catidad de veces. Cabe mencionar que a lo largo del proyecto no se profundizó en el cálculo de las raices de los polinomios de Legendre ni de los coeficeintes definidos en (1). En esté proyecto dichas raices y coeficientes se obtuvieron de la libreria scipy. Sería interesante ver la implementación númerica del cálculo de las raices y coeficientes de Legendre. Pues son fundamentales para este proyecto.

## Referencias

Burden, R. & Faires, J.. (2011). Numerical Differentiation and Integration. En Numerical Analysis(pp. 193-250). Boston: Cengage Learning.